pyquil/quil.py

```diff
@@ -558,43 +558,22 @@ def pop(self):
     def dagger(self, inv_dict=None, suffix="-INV"):
         """
         Creates the conjugate transpose of the Quil program. The program must not
-        contain any irreversible actions (measurement, control flow, qubit allocation).
+        contain only gate applications.
+
+        Note: the keyword arguments inv_dict and suffix are kept only
+        for backwards compatibility and have no effect.
 
         :return: The Quil program's inverse
         :rtype: Program
 
         """
-        if not self.is_protoquil():
-            raise ValueError("Program must be valid Protoquil")
-
-        daggered = Program()
-
-        for gate in self._defined_gates:
-            if inv_dict is None or gate.name not in inv_dict:
-                if gate.parameters:
-                    raise TypeError("Cannot auto define daggered version of parameterized gates")
-                daggered.defgate(gate.name + suffix, gate.matrix.T.conj())
-
-        for gate in reversed(self._instructions):
-            if gate.name in QUANTUM_GATES:
-                if gate.name == "S":
-                    daggered.inst(QUANTUM_GATES["PHASE"](-pi / 2, *gate.qubits))
-                elif gate.name == "T":
-                    daggered.inst(QUANTUM_GATES["RZ"](pi / 4, *gate.qubits))
-                elif gate.name == "ISWAP":
-                    daggered.inst(QUANTUM_GATES["PSWAP"](pi / 2, *gate.qubits))
-                else:
-                    negated_params = list(map(lambda x: -1 * x, gate.params))
-                    daggered.inst(QUANTUM_GATES[gate.name](*(negated_params + gate.qubits)))
-            else:
-                if inv_dict is None or gate.name not in inv_dict:
-                    gate_inv_name = gate.name + suffix
-                else:
-                    gate_inv_name = inv_dict[gate.name]
-
-                daggered.inst(Gate(gate_inv_name, gate.params, gate.qubits))
+        if not all(map(lambda instr: isinstance(instr, Gate), self._instructions)):
+            raise ValueError("Program to be daggered must be contain only gate applications")
 
-        return daggered
+        # This is a bit hacky. Gate.dagger() mutates the gate object,
```

Comment:
I'm confused by this comment. It seems like it _used_ to mutate, but now it's returning a copy. The comment makes it sound like it's not.

Reply:
The old way didn't mutate the original program, it went through each instruction and instantiated a new (dagg'r'd) version of it. The new way also doesn't mutate the original program.

My comment is that *`Gate.dagger()`* mutates. So I can't do `return Program([gate.dagger() for gate in original_program])` since that would mutate `original_program`.

Reply:
Why does Gate.dagger mutate though?

Reply:
Poor design choice by myself I guess :{